client/client-lib/src/query.rs

```diff
@@ -253,40 +238,44 @@ impl<R: Eq + Clone + Debug> QueryStrategy<R> for CurrentConsensus<R> {
                     self.existing_results.push((result, HashSet::from([peer])));
                 }
             }
-            FedResponse {
-                peer,
-                result: Err(error),
-            } => {
+            Err(error) => {
                 self.errors.insert(peer, error);
             }
         }
 
         for (result, peers) in &self.existing_results {
             if peers.len() >= self.required {
-                return QueryStep::Finished(Ok(result.clone()));
+                return QueryStep::Success(result.clone());
             }
         }
 
         if self.errors.len() >= self.required {
-            let (_, error) = self.errors.drain().next().expect("non-empty");
-            return QueryStep::Finished(Err(ApiError::RpcError(error)));
+            return QueryStep::Failure(mem::take(&mut self.errors));
         }
 
         QueryStep::Continue
     }
 }
 
 pub trait QueryStrategy<IR, OR = IR> {
-    fn process(&mut self, response: FedResponse<IR>) -> QueryStep<OR>;
+    fn process(&mut self, peer_id: PeerId, response: api::Result<IR>) -> QueryStep<OR>;
 }
 
 /// Results from the strategy handling a response from a peer
 ///
-/// `Retry` sending requests to some of the peers
-/// `Continue` awaiting and handling responses
-/// `Finished` return a final result
+/// Note that the implementation driving the [`QueryStrategy`] returning
+/// [`QueryStep`] is responsible from remembering and collecting errors
+/// for each peer.
+#[derive(Debug)]
```

Comment:
I didn't want to touch it, but change in the interface involved change in the way errors are handled, and that forced me to go over stuff here and tweak it a bit to fit.